mypy/meet.py

```diff
@@ -51,155 +56,246 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
+    upper bound.
 
-    If use_promotions is True, also consider type promotions (int and
-    float would only be overlapping if it's True).
+    This function is useful primarily when checking to see if two types are overlapping:
+    the algorithm to check if two unions are overlapping is fundamentally the same as
+    the algorithm for checking if two overloads are overlapping.
 
-    This does not consider multiple inheritance. For example, A and B in
-    the following example are not considered overlapping, even though
-    via C they can be overlapping:
+    Normalizing both kinds of types in the same way lets us reuse the same algorithm
+    for both.
+    """
+    if isinstance(typ, TypeVarType):
+        if len(typ.values) > 0:
+            return typ.values
+        else:
+            return [typ.upper_bound]
+    elif isinstance(typ, UnionType):
+        return typ.relevant_items()
+    elif isinstance(typ, Overloaded):
+        # Note: doing 'return typ.items()' makes mypy
+        # infer a too-specific return type of List[CallableType]
+        out = []  # type: List[Type]
+        out.extend(typ.items())
+        return out
+    else:
+        return [typ]
 
-        class A: ...
-        class B: ...
-        class C(A, B): ...
 
-    The rationale is that this case is usually very unlikely as multiple
-    inheritance is rare. Also, we can't reliably determine whether
-    multiple inheritance actually occurs somewhere in a program, due to
-    stub files hiding implementation details, dynamic loading etc.
+def is_overlapping_types(left: Type,
+                         right: Type,
+                         ignore_promotions: bool = False,
+                         prohibit_none_typevar_overlap: bool = False) -> bool:
+    """Can a value of type 'left' also be of type 'right' or vice-versa?
 
-    TODO: Don't consider callables always overlapping.
-    TODO: Don't consider type variables with values always overlapping.
+    If 'ignore_promotions' is True, we ignore promotions while checking for overlaps.
+    If 'prohibit_none_typevar_overlap' is True, we disallow None from overlapping with
+    TypeVars (in both strict-optional and non-strict-optional mode).
     """
-    # Any overlaps with everything
-    if isinstance(t, AnyType) or isinstance(s, AnyType):
+
+    def _is_overlapping_types(left: Type, right: Type) -> bool:
+        '''Encode the kind of overlapping check to perform.
+
+        This function mostly exists so we don't have to repeat keyword arguments everywhere.'''
+        return is_overlapping_types(
+            left, right,
+            ignore_promotions=ignore_promotions,
+            prohibit_none_typevar_overlap=prohibit_none_typevar_overlap)
+
+    # We should never encounter these types, but if we do, we handle
+    # them in the same way we handle 'Any'.
+    illegal_types = (UnboundType, PartialType, ErasedType, DeletedType)
```

Comment:
`PartialType` should be an `assert False, "Something bad about partial types"`, for other it is OK to be `Any`